synapse/storage/databases/main/events_worker.py

```diff
@@ -1617,21 +1618,31 @@ def _fetch_event_rows(
             # likely that some of these events may be for the same room/user combo, in
             # which case we don't need to do redundant queries
             to_check_set = set(to_check)
-            for room_and_user in to_check_set:
-                room_redactions_sql = "SELECT redacting_event_id, redact_end_ordering FROM room_ban_redactions WHERE room_id = ? and user_id = ?"
-                txn.execute(room_redactions_sql, room_and_user)
-
-                res = txn.fetchone()
-                # we have a redaction for a room, user_id combo - apply it to matching events
-                if not res:
-                    continue
+            room_redaction_sql = "SELECT room_id, user_id, redacting_event_id, redact_end_ordering FROM room_ban_redactions WHERE "
+            (
+                in_list_clause,
+                room_redaction_args,
+            ) = make_tuple_in_list_sql_clause(
+                self.database_engine, ("room_id", "user_id"), to_check_set
+            )
+            txn.execute(room_redaction_sql + in_list_clause, room_redaction_args)
+            res = txn.fetchall()
+            if not res:
+                continue
+            for (
+                returned_room_id,
+                returned_user_id,
+                redacting_event_id,
+                redact_end_ordering,
+            ) in res:
```

Comment:
You can iterate over `txn` to get the rows, rather than first reading them into a list and iterating over that. (This doesn't work if you then want to query the DB within the loop)

```suggestion
            for (
                returned_room_id,
                returned_user_id,
                redacting_event_id,
                redact_end_ordering,
            ) in txn:
```